# Aplicacion de fracciones continuadas para estimar $\pi$ y el GoldenRadio $\phi$.

In [ ]:
from sympy  import symbols
a = symbols('a0:6')
a

(a0, a1, a2, a3, a4, a5)

In [ ]:
x = symbols('x')

In [ ]:
def continued_fraction(depth):
    frac = a[0]
    for d in range(depth):
        frac = frac.replace( a[d], a[d] + x/a[d+1])
    return frac

Aproximemos $\pi$ usando fracciones continuadas.

In [ ]:
# podemos ver en Wikipedia la aproximacion de Pi en fracciones continuadas
# [3;7,15,1,292,1,1]
cf = continued_fraction(5)
cf

a0 + x/(a1 + x/(a2 + x/(a3 + x/(a4 + x/a5))))

In [ ]:
cfPI = cf.subs( [ (a(0), 3), (a[1],7), (a[2],13), (a[3],1), (a[4], 292), (a[5],1), (x,1) ])


ImportError: cannot import name 'subs' from 'sympy' (/usr/local/lib/python3.10/dist-packages/sympy/__init__.py)

In [ ]:
cfPI = cf.subs( [ (a[0], 3), (a[1], 7), (a[2],15), (a[3],1), (a[4],292),(a[5],1),(x,1)])


In [ ]:
print(cfPI)

104348/33215


In [ ]:
float(cfPI)

3.141592653921421

### Actividad 1.
Aproxime el GoldenRatio $(1 + \sqrt{5})/2$ con 30 coeficientes de $a_i$ en fracciones continuadas.

## Otras formas de evaluar

In [ ]:
import sympy as sp
display(cfPI) # fraccion exacta
print(float(cfPI)) # conversion a float
print(sp.N(cfPI)) # evaluacion numerica de SymPY

104348/33215

3.141592653921421
3.14159265392142


In [ ]:
## Como usar $n$ cifras decimales
sp.N(sp.pi, 30) # el pi de SymPy con 30 cifras decimales

3.14159265358979323846264338328

In [ ]:
sp.N(sp.pi, 1000) # se puede investigar hasta cuantas cifras maximo

3.14159265358979323846264338327950288419716939937510582097494459230781640628620899862803482534211706798214808651328230664709384460955058223172535940812848111745028410270193852110555964462294895493038196442881097566593344612847564823378678316527120190914564856692346034861045432664821339360726024914127372458700660631558817488152092096282925409171536436789259036001133053054882046652138414695194151160943305727036575959195309218611738193261179310511854807446237996274956735188575272489122793818301194912983367336244065664308602139494639522473719070217986094370277053921717629317675238467481846766940513200056812714526356082778577134275778960917363717872146844090122495343014654958537105079227968925892354201995611212902196086403441815981362977477130996051870721134999999837297804995105973173281609631859502445945534690830264252230825334468503526193118817101000313783875288658753320838142061717766914730359825349042875546873115956286388235378759375195778185778053217122680661300192787661119590921642019

In [ ]:
float(sp.pi)

3.141592653589793

In [ ]:
# evalf
sp.pi.evalf(30)

3.14159265358979323846264338328

In [ ]:
print("%.30f"%(sp.pi))

3.141592653589793115997963468544


### Actividad 2
Cual es la diferencia entre el print de arriba y el evalf?
Por que no dan lo mismo.

## Precision.

Algunas ideas de aca son tomadas del manual de SymPi.


Comentamos entre ```evalf```, ```N```, ```float```. Ilustramos
esto con un ejemplo tomado del manual.

La serie de Fibonacci: 1, 1, 2, 3, 5,.....

Se puede calcular exactamente que el termino $n$ de esta serie
se puede generar con la formula.

$$ \frac{\varphi^n - (1 - \varphi)^n}{\sqrt{5}} .$$
donde $\varphi$ es el GoldenRatio (algunas veces se usa la notacion $\phi$)



In [ ]:
from sympy import GoldenRatio, fibonacci
n=1000
a,b = (GoldenRatio**n - (1 - GoldenRatio)**n)/sp.sqrt(5), fibonacci(n)
display(float(a))
display(float(b))
float(a)- float(b)

4.3466557686937455e+208

4.3466557686937455e+208

0.0

In [ ]:
n=100
sp.N((  GoldenRatio**n - (1 - GoldenRatio)**n)/sp.sqrt(5) - fibonacci(n))

0.e-104

In [ ]:
n=1000
sp.N((  GoldenRatio**n - (1 - GoldenRatio)**n)/sp.sqrt(5) - fibonacci(n))

0.e+85

### Actividad 3
por que ese error tan grande? mientras al usar ```float``` el error no lo es tan grande.
Estudie el parametro ```maxn``` para mejorar la precision.


In [41]:
sp.N((  GoldenRatio**n - (1 - GoldenRatio)**n)/sp.sqrt(5) - fibonacci(n), maxn=500)

0.e-527

Podemos forzar el programa a que se aborte cuando la precision no es muy buena

In [42]:
sp.N((  GoldenRatio**n - (1 - GoldenRatio)**n)/sp.sqrt(5) - fibonacci(n), strict=True)

PrecisionExhausted: Failed to distinguish the expression: 

-43466557686937456435688527675040625802564660517371780402481729089536555417949051890403879840079255169295922593080322634775209689623239873322471161642996440906533187938298969649928516003704476137795166849228875 + sqrt(5)*(-(1 - GoldenRatio)**1000 + GoldenRatio**1000)/5

from zero. Try simplifying the input, using chop=True, or providing a higher maxn for evalf

In [43]:
sp.N((  GoldenRatio**n - (1 - GoldenRatio)**n)/sp.sqrt(5) - fibonacci(n), chop=True)

0.e+85

## ```nsimplify()```


In [44]:
from sympy import nsimplify # convierte decimal a fraccion
nsimplify(0.1)

1/10

In [48]:
nsimplify(6.28, [sp.pi], tolerance=0.001)

2*pi

In [52]:
nsimplify( sp.pi, tolerance=0.01)

22/7

In [53]:
nsimplify( sp.pi, tolerance=0.001)

355/113

In [54]:
# golden ratio
nsimplify( 4/(1 + sp.sqrt(5)), [GoldenRatio])

-2 + 2*GoldenRatio

In [55]:
# esta es del manual
from sympy import I
nsimplify(I**I, [sp.pi])

exp(-pi/2)

### Actividad 4.
Expliue el resultado anterior usando variable compleja.  Es decir
$$ \mathrm{i}^{\mathrm{i}} = \mathrm{e}^{-\pi/2}.$$

In [57]:
nsimplify(0.333333, rational='True') # conversion de decimal a fraccion

333333/1000000

### conversion de decimales a fracciones
Evaluacion de expresiones usando sumas.

$$0.333333 = \frac{3}{10} + \frac{3}{10^2} + \cdots + \frac{3}{10^6} = 3 \sum_{n=1}^6 \frac{1}{10^n} .$$


Sabemos que
$$ 1 + x + x^2 + \cdots + x^6 = \frac{1 - x^7}{1-x} .$$

Pasando el uno a restar


$$x + x^2 + \cdots + x^6 = \frac{1 -x^7}{1 -x} -1 = \frac{1-x^7 -1 + x}{1-x} = \frac{x (1 - x^6)}{1 -x}  .$$
En este caso $x=1/10$. Reemplazando


$$ 0.333333 = 3  \frac{10^{-1}(1 - 10^{-6})}{1 - 10^{-1}}
= 3 \frac{1 - 10^{-6}}{10 - 1} = \frac{10^6-1}{9 \times 10^6}
= 3 \frac{999999}{9 \times 1000000}  =  \frac{333333}{1000000}  .$$

Cual es la gracia, si ?


$$0.333333 \times \frac{1000000}{1000000} = \frac{333333}{1000000} .$$

0.4 a entero
multiplicando y dividiendo por 10 teneos

$$0.4 \frac{10}{10} = \frac{4}{10} .$$
pero hay que simplificar $2/5$.

In [58]:
nsimplify( 0.4, tolerance=0.1)

2/5

In [59]:
# hagamos la suma de arriba usando SimPy
from sympy import Sum
k=symbols('k')
s = 3*Sum(1/10**k, (k, 1, 6))
s

3*Sum(10**(-k), (k, 1, 6))

In [61]:
from sympy import simplify
simplify(s, rational="True")

333333/1000000

calcular el numero $1$

\begin{eqnarray}
1  &=& 1 -1 + 1 \\
1  &=& 1 -1 + 1 - 1 + 1 \\
1  &=& 1 -1 + 1 - 1 + 1 -1 + 1  \\
1  &=& 1 -1 + 1 - 1 + 1 -1 + 1 -1 + \cdots   \\
\end{eqnarray}
Es mas la serie $\sum_{n=0}^{\infty} (-1)^n$ no converge

$$ 1  &=& (1 -1) + (1 - 1) + (1 -1) + (1 -1) + \cdots  .$$

$$ 1 = 0 .$$

Godel: hay problemas que no se puede probar que son o no son solubles.

In [62]:
0**0

1

### Actividad 5.
Todo numero que no sea 0, elevado a la 0 es 1.
Ej, $10^0=1$.
$0$ elevado a cualquier potencia es 0,
pero $0^0=1$? por que?

Proxima clase comenzamos con **Algebra**